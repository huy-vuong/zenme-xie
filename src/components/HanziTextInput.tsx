import opencc from 'node-opencc';
import React, { useState } from 'react';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import FormControl from 'react-bootstrap/FormControl';
import styles from 'zenme-xie/components/HanziTextInput.module.scss';

enum CharSet {
  None,
  Traditional,
  Simplified,
}

export default function HanziTextInput({ text, setText }: HanziTextInputProp) {
  const [charSet, setCharSet] = useState(CharSet.None);
  return (
    <div className={styles.textInputGroup}>
      <FormControl
        type="text"
        size="lg"
        value={text}
        placeholder="你想寫什麼？"
        onChange={(e) => {
          setCharSet(CharSet.None);
          setText(e.target.value ?? '');
        }}
        className={styles.textInput}
      />
      <ButtonGroup size="lg">
        &nbsp;
        <Button
          variant={charSet === CharSet.Traditional ? 'light' : 'outline-light'}
          onClick={async () => {
            setCharSet(CharSet.Traditional);
            setText(opencc.simplifiedToTraditional(text));
          }}
        >
          T
        </Button>
        <Button
          variant={charSet === CharSet.Simplified ? 'light' : 'outline-light'}
          onClick={async () => {
            setCharSet(CharSet.Simplified);
            setText(opencc.traditionalToSimplified(text));
          }}
        >
          S
        </Button>
      </ButtonGroup>
    </div>
  );
}

interface HanziTextInputProp {
  text: string;
  setText: (text: string) => any;
}
