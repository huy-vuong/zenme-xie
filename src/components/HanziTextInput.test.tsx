import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import HanziTextInput from 'zenme-xie/components/HanziTextInput';

test('renders the given text', () => {
  const { getByDisplayValue, asFragment } = render(
    <HanziTextInput text="貓咪喵喵叫" setText={() => {}} />
  );
  getByDisplayValue('貓咪喵喵叫');
  expect(asFragment()).toMatchSnapshot();
});

test('converts the given text to Simplified when S is clicked', () => {
  const state = {
    text: '貓咪喵喵叫',
  };
  const { getByDisplayValue, getByText } = render(
    <HanziTextInput
      text={state.text}
      setText={(text) => {
        state.text = text;
      }}
    />
  );
  getByDisplayValue('貓咪喵喵叫');
  fireEvent.click(getByText('S'));
  expect(state.text).toBe('猫咪喵喵叫');
});

test('converts the given text to Traditional when T is clicked', () => {
  const state = {
    text: '猫咪喵喵叫',
  };
  const { getByDisplayValue, getByText } = render(
    <HanziTextInput
      text={state.text}
      setText={(text) => {
        state.text = text;
      }}
    />
  );
  getByDisplayValue('猫咪喵喵叫');
  fireEvent.click(getByText('T'));
  expect(state.text).toBe('貓咪喵喵叫');
});
