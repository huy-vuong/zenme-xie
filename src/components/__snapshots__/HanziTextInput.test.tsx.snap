// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the given text 1`] = `
<DocumentFragment>
  <div
    class="textInputGroup"
  >
    <input
      class="textInput form-control form-control-lg"
      placeholder="你想寫什麼？"
      type="text"
      value="貓咪喵喵叫"
    />
    <div
      class="btn-group btn-group-lg"
      role="group"
    >
       
      <button
        class="btn btn-outline-light"
        type="button"
      >
        T
      </button>
      <button
        class="btn btn-outline-light"
        type="button"
      >
        S
      </button>
    </div>
  </div>
</DocumentFragment>
`;
